File: 2020/python/day02.py
import unittest


def parse(line):
    r, c, password = line.split(' ')
    rs, re = r.split('-')
    c, *tail = c.split(':')
    return (password, c, (int(rs), int(re)))


def char_count(password, c):
    return password.count(c)


def match_policy(password, c, occurences):
    min_occ, max_occ = occurences
    count = char_count(password, c)
    return min_occ <= count <= max_occ


def run(entries):
    result = 0
    for entry in entries:
        password, c, occurences = entry
        if (match_policy(password, c, occurences)):
            result += 1
    return result


def part1():
    print(run(read_input()))


def char_pos(password, c, pos):
    return password[pos - 1] == c


def match_policy2(password, c, positions):
    return char_pos(password, c, positions[0]) ^ char_pos(
        password, c, positions[1])


def run2(entries):
    result = 0
    for entry in entries:
        password, c, occurences = entry
        if (match_policy2(password, c, occurences)):
            result += 1
    return result


def part2():
    print(run2(read_input()))


def read_input():
    with open('../inputs/input02') as input:
        return [parse(i) for i in input if not i.isspace()]


class TestPart1(unittest.TestCase):
    def test_parse(self):
        self.assertEqual(parse('1-3 a: abcde'), ('abcde', 'a', (1, 3)))

    def test_char_count(self):
        self.assertEqual(char_count('abcde', 'a'), 1)
        self.assertEqual(char_count('cdefg', 'b'), 0)
        self.assertEqual(char_count('ccccccccc', 'c'), 9)

    def test_match_policy(self):
        self.assertTrue(match_policy('abcde', 'a', (1, 3)))
        self.assertFalse(match_policy('cdefg', 'b', (1, 3)))
        self.assertTrue(match_policy('ccccccccc', 'c', (2, 9)))

    def test_run(self):
        self.assertEqual(
            run([('abcde', 'a', (1, 3)), ('cdefg', 'b', (1, 3)),
                 ('ccccccccc', 'c', (2, 9))]), 2)


class TestPart2(unittest.TestCase):
    def test_char_pos(self):
        self.assertTrue(char_pos('abcde', 'a', 1))
        self.assertFalse(char_pos('cdefg', 'b', 1))
        self.assertTrue(char_pos('ccccccccc', 'c', 2))

    def test_match_policy2(self):
        self.assertTrue(match_policy2('abcde', 'a', (1, 3)))
        self.assertFalse(match_policy2('cdefg', 'b', (1, 3)))
        self.assertFalse(match_policy2('ccccccccc', 'c', (2, 9)))

    def test_run2(self):
        self.assertEqual(
            run2([('abcde', 'a', (1, 3)), ('cdefg', 'b', (1, 3)),
                  ('ccccccccc', 'c', (2, 9))]), 1)


if __name__ == '__main__':
    part1()
    part2()
    unittest.main()
    pass

File: 2020/rust-sln/src/day02.rs
use std::error::Error;
use std::fmt;
use std::num::ParseIntError;
use std::str::FromStr;

fn main() {
    println!("Part1 {}", part1());
    println!("Part2 {}", part2());
}

#[derive(Debug, Clone)]
struct ParseEntryError;
impl Error for ParseEntryError {}
impl fmt::Display for ParseEntryError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "Could not parse to Entry")
    }
}
impl From<ParseIntError> for ParseEntryError {
    fn from(err: ParseIntError) -> ParseEntryError {
        ParseEntryError
    }
}

#[derive(Debug, Default, PartialEq)]
struct Entry {
    c: char,
    password: String,
    range: (usize, usize),
}

impl FromStr for Entry {
    type Err = ParseEntryError;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let tokens = s.split_whitespace().collect::<Vec<_>>();
        if tokens.len() < 3 {
            return Err(ParseEntryError);
        }
        let min_max = tokens[0].split("-").collect::<Vec<_>>();
        let c = tokens[1].split(":").take(1).collect::<Vec<_>>();
        let password = tokens[2];

        let min = min_max[0].parse::<usize>()?;
        let max = min_max[1].parse::<usize>()?;
        let c = c[0].chars().next().unwrap();
        Ok(Entry {
            password: password.to_string(),
            c,
            range: (min, max),
        })
    }
}

fn char_count(password: &str, c: char) -> usize {
    password.matches(c).count()
}

fn policy_minmax(entry: &Entry) -> bool {
    let min = entry.range.0;
    let max = entry.range.1;
    let count = char_count(&entry.password, entry.c);
    min <= count && count <= max
}

fn match_policy(entry: &Entry, policy: fn(&Entry) -> bool) -> bool {
    return policy(entry);
}

fn run(entries: &Vec<Entry>, policy: fn(&Entry) -> bool) -> usize {
    entries
        .iter()
        .filter(|entry| match_policy(entry, policy))
        .count()
}

fn read_input() -> Vec<Entry> {
    include_str!("../../inputs/input02")
        .lines()
        .flat_map(|line| Entry::from_str(line))
        .collect::<Vec<_>>()
}

fn part1() -> usize {
    run(&read_input(), policy_minmax)
}

fn char_pos(password: &str, c: char, pos: usize) -> bool {
    if let Some(tc) = password.chars().nth(pos - 1) {
        return tc == c;
    }
    false
}

fn policy_position(entry: &Entry) -> bool {
    char_pos(&entry.password, entry.c, entry.range.0)
        ^ char_pos(&entry.password, entry.c, entry.range.1)
}

fn part2() -> usize {
    run(&read_input(), policy_position)
}

mod tests {
    use super::*;

    #[test]
    fn test_parse() {
        assert_eq!(
            Entry::from_str("1-3 a: abcde").unwrap(),
            Entry {
                c: 'a',
                password: "abcde".into(),
                range: (1, 3)
            }
        )
    }

    #[test]
    fn test_char_count() {
        assert_eq!(char_count("abcde", 'a'), 1);
        assert_eq!(char_count("cdefg", 'b'), 0);
        assert_eq!(char_count("ccccccccc", 'c'), 9);
    }

    #[test]
    fn test_policy_minmax() {
        assert!(policy_minmax(&Entry {
            password: "abcde".into(),
            c: 'a',
            range: (1, 3)
        }));
    }

    #[test]
    fn test_run() {
        let list = vec![
            Entry {
                password: "abcde".into(),
                c: 'a',
                range: (1, 3),
            },
            Entry {
                password: "cdefg".into(),
                c: 'b',
                range: (1, 3),
            },
            Entry {
                password: "ccccccccc".into(),
                c: 'c',
                range: (2, 9),
            },
        ];
        assert_eq!(run(&list, policy_minmax), 2);
    }

    #[test]
    fn test_policy_position() {
        assert!(policy_position(&Entry {
            password: "abcde".into(),
            c: 'a',
            range: (1, 3)
        }));
    }
}
